src/Functions/tupleToNameValuePairs.cpp

```diff
@@ -0,0 +1,132 @@
+{
+
+/** Transform a named tuple into an array of pairs, where the first element
+  * of the pair corresponds to the tuple field name and the second one to the
+  * tuple value.
+  */
+class FunctionTupleToNameValuePairs : public IFunction
+{
+public:
+    static constexpr auto name = "tupleToNameValuePairs";
+    static FunctionPtr create(ContextPtr)
+    {
+        return std::make_shared<FunctionTupleToNameValuePairs>();
+    }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    size_t getNumberOfArguments() const override
+    {
+        return 1;
+    }
+
+    bool useDefaultImplementationForConstants() const override
+    {
+        return true;
+    }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        // get the type of all the fields in the tuple
+        const IDataType * col = arguments[0].type.get();
+        const DataTypeTuple * tuple = checkAndGetDataType<DataTypeTuple>(col);
+
+        if (!tuple)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                            "First argument for function {} must be a tuple.",
+                            getName());
+
+        const auto & elementTypes = tuple->getElements();
+
+        if (elementTypes.empty())
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                            "The argument tuple for function {} must not be empty.",
+                            getName());
+
+        const auto & firstElementType = elementTypes[0];
+
+        bool allValueTypesEqual = std::all_of(elementTypes.begin() + 1,
+                                              elementTypes.end(),
+                                              [&](const auto &other)
+                                              {
+                                                  return firstElementType->equals(*other);
+                                              });
+
+        if (!allValueTypesEqual)
+        {
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                            "The argument tuple for function {} must contain just one type.",
+                            getName());
+        }
+
+        DataTypePtr tupleNameType = std::make_shared<DataTypeString>();
+        DataTypes itemDataTypes ={tupleNameType,
+                                  firstElementType};
+
+        auto itemDataType = std::make_shared<DataTypeTuple>(itemDataTypes);
+
+        return std::make_shared<DataTypeArray>(itemDataType);
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t /*input_rows_count*/) const override
+    {
+        const IColumn * tuple_col = arguments[0].column.get();
+        const DataTypeTuple * tuple = checkAndGetDataType<DataTypeTuple>(arguments[0].type.get());
+        auto * tuple_col_concrete = assert_cast<const ColumnTuple*>(tuple_col);
+
+        MutableColumnPtr keys = ColumnString::create();
```

Comment:
Better to put auto here, of specify ColumnString pointer type, it is important because on line 114
```
keys->insertData(key.data(), key.size());
```
can be devirtualized.

Reply:
Ha that's a good one